future-apps/payroll/contracts/Payroll.sol

```diff
@@ -0,0 +1,569 @@
+            // check token is allowed
+            require(allowedTokens[tokens[i]]);
+            // set distribution
+            employee.allocation[tokens[i]] = distribution[i];
+            sum = sum.add(distribution[i]);
+        }
+        require(sum == 100);
+
+        employee.lastAllocation = getTimestamp();
+    }
+
+    /**
+     * @dev Get payment proportion for a token and an employee (the caller)
+     * @notice Get payment proportion for a token and an employee (the caller)
+     * @param token The token address
+     */
+    function getAllocation(address token) public view returns(uint256 allocation) {
+        var employee = employees[employeeIds[msg.sender]];
+        // check that employee exists (and matches)
+        require(employee.accountAddress == msg.sender);
+
+        allocation = employee.allocation[token];
+    }
+
+    /**
+     * @dev payday To withdraw monthly payment by employee (the caller)
+     * @notice Only callable once a month. Assumed token has these standard checks implemented:
+     *         https://theethereum.wiki/w/index.php/ERC20_Token_Standard#How_Does_A_Token_Contract_Work.3F
+     */
+    function payday() public {
+        var employee = employees[employeeIds[msg.sender]];
+        // check that employee exists (and matches)
+        require(employee.accountAddress == msg.sender);
+        // get time that has gone by (seconds)
+        uint256 time = getTimestamp().sub(employee.lastPayroll);
+        require(time > 0);
+
+        // loop over allowed tokens
+        bool somethingPaid = false;
+        for (uint i = 0; i < allowedTokensArray.length; i++) {
+            var token = allowedTokensArray[i];
+            if (employee.allocation[token] == 0)
+                continue;
+            require(checkExchangeRate(token));
+            // salary converted to token and applied allocation percentage
+            uint256 tokenAmount = employee.denominationTokenSalary
+                .mul(exchangeRates[token]).mul(employee.allocation[token]) / 100;
+            tokenAmount = tokenAmount.mul(time);
+            ERC20 tokenContract = ERC20(token);
+            finance.newPayment(
+                tokenContract,
+                msg.sender,
+                tokenAmount,
+                0,
+                0,
+                1,
+                ""
+            );
+            LogSendPayroll(msg.sender, token, tokenAmount);
+            somethingPaid = true;
+        }
+        require(somethingPaid);
+        // finally update last payroll date
+        employee.lastPayroll = getTimestamp();
+    }
+
+    /**
+     * @dev Change employee account address. To be called by Employer.
+     * @notice Change employee account address
+     * @param employeeId Employee's identifier
+     * @param newAddress New address to receive payments
+     */
+    function changeAddressByOwner(uint256 employeeId, address newAddress) public auth(EMPLOYER_ROLE) {
```

Comment:
I don't like this because it allows the employer to 'steal' funds from an employee that hasn't done payday() in a while

Reply:
Well, yes, but then should we remove the account address parameter from the addEmployee function and let the employee to set it? This would complicate too much the flow, I think. Anyway, I agree with removing this function, as employee can do it so it's not necessary.